File: macos/Scanner/Sources/Scanner/RulePack/RulePackEvaluator.swift
public struct RulePackEvaluator: Sendable {
    public init() {}

    public func evaluate(
        facts: ScanFacts,
        packs: [RulePack],
        existingFindings: [Finding]
    ) -> [Finding] {
        var findings = existingFindings

        for pack in packs {
            // 1. Custom rules → new findings
            for rule in pack.rules {
                findings.append(contentsOf: matchRule(rule, against: facts))
            }

            // 2. Overrides — replace severity on matched rule IDs
            let overrideMap = Dictionary(
                uniqueKeysWithValues: pack.overrides.map { ($0.id, $0.severity) }
            )
            if !overrideMap.isEmpty {
                findings = findings.map { finding in
                    guard let newSeverity = overrideMap[finding.ruleId] else {
                        return finding
                    }
                    return Finding(
                        ruleId: finding.ruleId,
                        severity: newSeverity,
                        app: finding.app,
                        serverName: finding.serverName,
                        extensionId: finding.extensionId,
                        affectedPath: finding.affectedPath,
                        maskedValue: finding.maskedValue,
                        title: finding.title,
                        explanation: finding.explanation,
                        recommendation: finding.recommendation,
                        escalationReason: finding.escalationReason
                    )
                }
            }
        }

        return findings
    }

    // MARK: - Match

    private func matchRule(_ rule: PackRule, against facts: ScanFacts) -> [Finding] {
        switch rule.match.fact {
        case .mcpServer:
            return facts.mcpServers
                .filter { matches(server: $0, match: rule.match) }
                .map { server in
                    Finding(
                        ruleId: rule.id,
                        severity: rule.severity,
                        app: server.appId,
                        serverName: server.name,
                        affectedPath: server.configPath,
                        title: rule.title,
                        explanation: rule.explanation,
                        recommendation: rule.recommendation
                    )
                }

        case .setting:
            return facts.settings
                .filter { matches(setting: $0, match: rule.match) }
                .map { setting in
                    Finding(
                        ruleId: rule.id,
                        severity: rule.severity,
                        app: setting.appId,
                        affectedPath: setting.configPath,
                        title: rule.title,
                        explanation: rule.explanation,
                        recommendation: rule.recommendation
                    )
                }

        case .extension:
            return facts.extensions
                .filter { matches(ext: $0, match: rule.match) }
                .map { ext in
                    Finding(
                        ruleId: rule.id,
                        severity: rule.severity,
                        app: ext.appId,
                        extensionId: ext.extensionId,
                        title: rule.title,
                        explanation: rule.explanation,
                        recommendation: rule.recommendation
                    )
                }

        case .authFile:
            return facts.authFiles
                .filter { matches(auth: $0, match: rule.match) }
                .map { auth in
                    Finding(
                        ruleId: rule.id,
                        severity: rule.severity,
                        app: auth.appId,
                        affectedPath: auth.filePath,
                        title: rule.title,
                        explanation: rule.explanation,
                        recommendation: rule.recommendation
                    )
                }
        }
    }

    // MARK: - Predicate helpers

    private func matches(server: McpServerFact, match: PackRuleMatch) -> Bool {
        if let v = match.commandEquals, server.commandName != v { return false }
        if let v = match.commandContains, !server.commandName.contains(v) { return false }
        if let v = match.nameContains, !server.name.contains(v) { return false }
        if let v = match.argsContain, !server.args.contains(where: { $0.contains(v) }) { return false }
        if let v = match.app, server.appId != v { return false }
        return true
    }

    private func matches(setting: SettingFact, match: PackRuleMatch) -> Bool {
        if let v = match.key, setting.key != v { return false }
        if let v = match.valueEquals, setting.stringValue != v { return false }
        if let v = match.app, setting.appId != v { return false }
        return true
    }

    private func matches(ext: ExtensionFact, match: PackRuleMatch) -> Bool {
        if let v = match.extensionIdContains, !ext.extensionId.contains(v) { return false }
        if let v = match.app, ext.appId != v { return false }
        return true
    }

    private func matches(auth: AuthFileFact, match: PackRuleMatch) -> Bool {
        if let v = match.app, auth.appId != v { return false }
        return true
    }
}

File: windows/Scanner/RulePack/RulePackEvaluator.cs
using AIExposureScanner.Scanner.Models;

namespace AIExposureScanner.Scanner.RulePacks;

public sealed class RulePackEvaluator
{
    public IReadOnlyList<Finding> Evaluate(
        ScanFacts facts,
        IReadOnlyList<RulePack> packs,
        IReadOnlyList<Finding> existing)
    {
        var findings = new List<Finding>(existing);

        foreach (var pack in packs)
        {
            // 1. Custom rules → new findings
            foreach (var rule in pack.Rules)
                findings.AddRange(MatchRule(rule, facts));

            // 2. Overrides — replace severity on matched rule IDs
            var overrideMap = pack.Overrides.ToDictionary(o => o.Id, o => o.Severity, StringComparer.Ordinal);
            if (overrideMap.Count > 0)
            {
                for (var i = 0; i < findings.Count; i++)
                {
                    var f = findings[i];
                    if (overrideMap.TryGetValue(f.RuleId, out var newSev)
                        && Enum.TryParse<Severity>(newSev, ignoreCase: true, out var parsedSev))
                    {
                        findings[i] = f with { Severity = parsedSev };
                    }
                }
            }
        }

        return findings;
    }

    // MARK: - Match

    private static IEnumerable<Finding> MatchRule(PackRule rule, ScanFacts facts)
    {
        if (!Enum.TryParse<Severity>(rule.Severity, ignoreCase: true, out var severity))
            yield break;

        switch (rule.Match.Fact.ToLowerInvariant())
        {
            case "mcp_server":
                foreach (var s in facts.McpServers.Where(s => MatchesServer(s, rule.Match)))
                    yield return new Finding(rule.Id, severity, s.AppId,
                        ServerName: s.Name, AffectedPath: s.ConfigPath,
                        Title: rule.Title, Explanation: rule.Explanation, Recommendation: rule.Recommendation);
                break;

            case "setting":
                foreach (var s in facts.Settings.Where(s => MatchesSetting(s, rule.Match)))
                    yield return new Finding(rule.Id, severity, s.AppId,
                        AffectedPath: s.ConfigPath,
                        Title: rule.Title, Explanation: rule.Explanation, Recommendation: rule.Recommendation);
                break;

            case "extension":
                foreach (var e in facts.Extensions.Where(e => MatchesExtension(e, rule.Match)))
                    yield return new Finding(rule.Id, severity, e.AppId,
                        ExtensionId: e.ExtensionId,
                        Title: rule.Title, Explanation: rule.Explanation, Recommendation: rule.Recommendation);
                break;

            case "auth_file":
                foreach (var a in facts.AuthFiles.Where(a => MatchesAuth(a, rule.Match)))
                    yield return new Finding(rule.Id, severity, a.AppId,
                        AffectedPath: a.FilePath,
                        Title: rule.Title, Explanation: rule.Explanation, Recommendation: rule.Recommendation);
                break;
        }
    }

    // MARK: - Predicate helpers

    private static bool MatchesServer(McpServerFact s, PackRuleMatch m)
    {
        if (m.CommandEquals is not null && s.Command != m.CommandEquals) return false;
        if (m.CommandContains is not null && !s.Command.Contains(m.CommandContains)) return false;
        if (m.NameContains is not null && !s.Name.Contains(m.NameContains)) return false;
        if (m.ArgsContain is not null && !s.Args.Any(a => a.Contains(m.ArgsContain))) return false;
        if (m.App is not null && s.AppId != m.App) return false;
        return true;
    }

    private static bool MatchesSetting(SettingFact s, PackRuleMatch m)
    {
        if (m.Key is not null && s.Key != m.Key) return false;
        if (m.ValueEquals is not null && s.StringValue != m.ValueEquals) return false;
        if (m.App is not null && s.AppId != m.App) return false;
        return true;
    }

    private static bool MatchesExtension(ExtensionFact e, PackRuleMatch m)
    {
        if (m.ExtensionIdContains is not null && !e.ExtensionId.Contains(m.ExtensionIdContains)) return false;
        if (m.App is not null && e.AppId != m.App) return false;
        return true;
    }

    private static bool MatchesAuth(AuthFileFact a, PackRuleMatch m)
    {
        if (m.App is not null && a.AppId != m.App) return false;
        return true;
    }
}
